apm-lambda-extension/extension/apm_server_test.go

```diff
@@ -94,13 +90,204 @@ func TestPostToApmServerDataNotCompressed(t *testing.T) {
+	assert.InDelta(t, val2, float64(4), 0.1*4)
+
+	ApmServerTransportState.ReconnectionCount = 3
+	val3 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val3, float64(9), 0.1*9)
+
+	ApmServerTransportState.ReconnectionCount = 4
+	val4 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val4, float64(16), 0.1*16)
+
+	ApmServerTransportState.ReconnectionCount = 5
+	val5 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val5, float64(25), 0.1*25)
+
+	ApmServerTransportState.ReconnectionCount = 6
+	val6 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val6, float64(36), 0.1*36)
+
+	ApmServerTransportState.ReconnectionCount = 7
+	val7 := computeGracePeriod().Seconds()
+	assert.InDelta(t, val7, float64(36), 0.1*36)
+}
+
+func TestEnterBackoffFromHealthy(t *testing.T) {
+	SetApmServerTransportState(Healthy, context.Background())
+	// Compress the data
+	pr, pw := io.Pipe()
+	gw, _ := gzip.NewWriterLevel(pw, gzip.BestSpeed)
+	go func() {
+		gw.Write([]byte(""))
+		gw.Close()
+		pw.Close()
+	}()
+
+	// Create AgentData struct with compressed data
+	data, _ := ioutil.ReadAll(pr)
+	agentData := AgentData{Data: data, ContentEncoding: "gzip"}
+
+	// Create apm server and handler
+	apmServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, _ := ioutil.ReadAll(r.Body)
+		assert.Equal(t, string(data), string(bytes))
+		assert.Equal(t, "gzip", r.Header.Get("Content-Encoding"))
+		w.Write([]byte(`{"foo": "bar"}`))
+	}))
+	apmServer.Close() // Close the APM server early so that POST requests fail and that backoff is enabled
+
+	config := extensionConfig{
+		apmServerUrl: apmServer.URL + "/",
+	}
+
+	PostToApmServer(apmServer.Client(), agentData, &config, context.Background())
+	// No way to know for sure if failing or pending (0 sec grace period)
+	assert.True(t, ApmServerTransportState.Status != Healthy)
+	assert.Equal(t, ApmServerTransportState.ReconnectionCount, 0)
+}
+
+func TestEnterBackoffFromFailing(t *testing.T) {
+	SetApmServerTransportState(Healthy, context.Background())
+	SetApmServerTransportState(Failing, context.Background())
+	for {
+		if IsTransportStatusHealthyOrPending() {
+			break
+		}
+	}
+	assert.Equal(t, ApmServerTransportState.Status, Pending)
+
+	// Compress the data
+	pr, pw := io.Pipe()
+	gw, _ := gzip.NewWriterLevel(pw, gzip.BestSpeed)
+	go func() {
+		gw.Write([]byte(""))
+		gw.Close()
+		pw.Close()
+	}()
+
+	// Create AgentData struct with compressed data
+	data, _ := ioutil.ReadAll(pr)
+	agentData := AgentData{Data: data, ContentEncoding: "gzip"}
+
+	// Create apm server and handler
+	apmServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, _ := ioutil.ReadAll(r.Body)
+		assert.Equal(t, string(data), string(bytes))
+		assert.Equal(t, "gzip", r.Header.Get("Content-Encoding"))
+		w.Write([]byte(`{"foo": "bar"}`))
+	}))
+	apmServer.Close() // Close the APM server early so that POST requests fail and that backoff is enabled
```

Comment:
I'd put the comment above the line where the server is closed

Reply:
Fixed in 384b913